plutus-core/untyped-plutus-core/src/UntypedPlutusCore/NewParser.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeOperators       #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | Parser for untyped Plutus Core using megaparsec, as in Plutus IR.
+
+module UntypedPlutusCore.NewParser
+    ( topSourcePos
+    , parse
+    , parseQuoted
+    , term
+    , program
+    , parseTerm
+    , parseProgram
+    , applyProgram
+    , parseScoped
+    , Parser
+    , ParseError (..)
+    , Error
+    , SourcePos
+    ) where
+
+import           Prelude                         hiding (fail)
+
+import           Control.Monad.Except            ((<=<))
+import           Control.Monad.State             (StateT)
+
+import qualified NewUntypedPlutusCore            as UPLC
+import qualified PlutusCore                      as PLC
+import qualified PlutusCore.Parsable             as PLC
+import           PlutusPrelude                   (Pretty, through)
+import           Text.Megaparsec                 hiding (ParseError, State, parse)
+import           UntypedPlutusCore.Check.Uniques (checkProgram)
+import           UntypedPlutusCore.Rename        (Rename (rename))
+
+import           Data.ByteString.Lazy            (ByteString)
+import           Data.ByteString.Lazy.Internal   (unpackChars)
+import qualified Data.Text                       as T
+import           PlutusCore.ParserCommon
+
+-- A small type wrapper for parsers that are parametric in the type of term they parse
+type Parametric name uni fun
```

Comment:
Unused? Although I guess you could use it for the parsers in this file if you wanted to use the `TermLike` functions.

Reply:
Ah I see, that's what the PIR one does. Up to you: not sure how interesting the polymorphism is. If we did do it that way you could put them in the common file.

Reply:
but this is for UPLC's `TermLike`. I could parameterised it but I don't think it's worth it?

Reply:
yeah, we can revisit the idea of trying to abstract over the parsers with `TermLike` later if we want